ql/lib/codeql/kaleidoscope/Cfg.qll

```diff
@@ -0,0 +1,254 @@
+private import codeql.kaleidoscope.Ast
+private import codeql.controlflow.Cfg as CfgShared
+private import codeql.Locations
+private import codeql.kaleidoscope.controlflow.BasicBlocks
+
+private module Completion {
+  private newtype TCompletion =
+    TSimpleCompletion() or
+    TBooleanCompletion(boolean b) { b in [false, true] } or
+    TReturnCompletion()
+
+  abstract class Completion extends TCompletion {
+    abstract string toString();
+
+    predicate isValidForSpecific(AstNode e) { none() }
+
+    predicate isValidFor(AstNode e) { this.isValidForSpecific(e) }
+
+    abstract SuccessorType getAMatchingSuccessorType();
+  }
+
+  class SimpleCompletion extends Completion, TSimpleCompletion {
+    override string toString() { result = "SimpleCompletion" }
+
+    override predicate isValidFor(AstNode e) { not any(Completion c).isValidForSpecific(e) }
+
+    override NormalSuccessor getAMatchingSuccessorType() { any() }
+  }
+
+  class BooleanCompletion extends Completion, TBooleanCompletion {
+    boolean value;
+
+    BooleanCompletion() { this = TBooleanCompletion(value) }
+
+    override string toString() { result = "BooleanCompletion(" + value + ")" }
+
+    override predicate isValidForSpecific(AstNode e) {
```

Comment:
How about boolean expressions involving short-circuiting?

Reply:
I don't think kaleidoscope has that feature.